include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,487 @@
+
+    /*!\name Iterators
+     * \{
+     */
+    /*!\brief Returns an iterator to the first element of the container.
+     * \returns Iterator to the first element.
+     *
+     * If the container is empty, the returned iterator will be equal to end().
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * No-throw guarantee.
+     */
+    iterator begin() noexcept
+    {
+        return iterator{*this};
+    }
+
+    //!\copydoc begin()
+    const_iterator begin() const noexcept
+    {
+        return const_iterator{*this};
+    }
+
+    //!\copydoc begin()
+    const_iterator cbegin() const noexcept
+    {
+        return const_iterator{*this};
+    }
+
+    /*!\brief Returns an iterator to the element following the last element of the decorator.
+     * \returns Iterator to the behind last element.
+     *
+     * This element acts as a placeholder; attempting to dereference it results in undefined behaviour.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * No-throw guarantee.
+     */
+    iterator end() noexcept
+    {
+        return iterator{*this, size()};
+    }
+
+    //!\copydoc end()
+    const_iterator end() const noexcept
+    {
+        return const_iterator{*this, size()};
+    }
+
+    //!\copydoc end()
+    const_iterator cend() const noexcept
+    {
+        return const_iterator{*this, size()};
+    }
+    //!\}
+
+    /*!\name Element access
+     * \{
+     */
+    /*!\brief Return the i-th element as a reference.
+     * \param i The element to retrieve.
+     * \throws std::out_of_range If you access an element behind the last.
+     * \returns A reference of the gapped alphabet type .
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Strong exception guarantee (never modifies data)..
+     */
+    reference at(size_type const i)
+    {
+        //TODO add SEQAN_UNLIKELY
+        if (i >= size())
+            throw std::out_of_range{"Trying to access element behind the last in gap_decorator."};
+        return (*this)[i];
+    }
+
+    //!\copydoc at()
+    const_reference at(size_type const i) const
+    {
+        //TODO add SEQAN_UNLIKELY
+        if (i >= size())
+            throw std::out_of_range{"Trying to access element behind the last in gap_decorator."};
+        return (*this)[i];
+    }
+
+    //!\copydoc at()
+    constexpr reference operator[](size_type const i) const
+    {
+        assert(i < size());
+        // case 1: no gaps
+        if (!anchors.size())
+            return value_type((*sequence)[i]);
+        // case 2: gaps
+        anchor_set_iterator it = anchors.lower_bound(gap_t{i, _});
+        if (it == anchors.end())
+            it = std::prev(it);
+
+        size_type acc{0}, gap_len{0};
+        if ((*it).first <= i || (it != anchors.begin() && (*(std::prev(it))).first <= i))
+        {
+            if ((*it).first > i)
```

Comment:
As I understand it, you use lower_bound to find the first pair where `first` greater or equal to `i` but then you decrement if it is the end or if `first` is greater. So what you are actually searching for is the last element that is smaller or equal to `i` correct? If so I think it is more efficient if you use upper_bound and always `std::prev` the result. What do you think? You would save at least two if clauses if I am correct, but I am not sure if I got it right.

Reply:
Makes no difference and I still would have to check for it not being `begin()` before applying `prev`.

Reply:
Yes you are right. 
maybe you are still testing things more than necessary. Copying your code, I think it can be reduced to this:
```cpp        
        // case 2: gaps
        anchor_set_iterator it = anchors.upper_bound(gap_t{i, _});

        if (it == anchors.begin())
            return value_type((*sequence)[i]); // since no gaps happen before i
        
        it = std::prev(it); // here you can be sure that (*it).first <= i by defnition of upper_bound. No need to test.
        
        size_type gap_len{(*it).second};
        if (it != anchors.begin())
             gap_len -= (*(std::prev(it, 1))).second;

        if (i < (*it).first + gap_len)
            return gap::GAP;
        else
            return value_type((*sequence)[i - (*it).second]);

```
3 instead of 5 if clauses,
3 instead of 8 comparisons (counting all `</>/<=/>=/==/!=`)

Reply:
Yes, this would be a great de-obfuscation, I am testing this now in my benchmark branch with some operational tests that I have there.

Reply:
Great, the functional, randomized benchmark is running through. Though, I don't test semantics, i.e. sequence identity. This will be covered by the next unit test and fixed in case errors occur.